### ISSUE TITLE
Problem with max_weights in HERC

### ISSUE THREAD

**Issue by myllo** [skfolio/skfolio#91]
First, thank you for this library. I do appreciate that you have created skfolio and made it available for any interested investor. 

I like the idea of HERC, which resonates to my own experiences on the markets. I aim for reasonably high cumulative return while valuing a smooth return curve for the portfolio. 

However, using HERC as such, it seems quite often to lead to portfolio compositions which may be computationally correct, but very unbalanced.  As a remedy, I have tried to use weight constraints to hinder too large deviation from naïve (1/n) weighting in portfolio composition.

Here I have run into problems. Setting max_weights does not seem to have effect I thought it would. It does not seem to cap the weight of any single asset in the portfolio, as I’d like to see.

Using your example of Hierarchical Equal Risk Contribution – CdaR as a reference,
there are 16 assets in the portfolio.

If I have code like this

```
# The relevant parts only

weight_leeway = 0.05
max_weights = float(1/len(assets) + weight_leeway) # 0.1125
transaction_costs = float(0.004/(21*4))

assets = [
'BRK-B',
 'CME',
 'INVE-B.ST',
 'V',
 'NDA-FI.HE',
 'MUV2.DE',
 'MSFT',
 'NOVO-B.CO',
 'VVSM.DE',
 'TW',
 'FICO',
 'ALV.DE',
 'WRT1V.HE',
 'SAP.DE',
 'GOOGL',
 'BWXT']

# download the data from yf, startdate=2020-01-01 end_date=2024-10-21 and remove the blanks (NaN)

X = prices_to_returns(prices)
X_train, X_test = train_test_split(X, test_size=0.33, shuffle=False)

model3 = HierarchicalEqualRiskContribution(
    risk_measure=RiskMeasure.CDAR,
    distance_estimator=KendallDistance(absolute=True),
    max_weights=max_weights,
    transaction_costs=transaction_costs,
    portfolio_params=dict(name="HERC-CDaR-Ward-Kendal")
)
model3.fit(X_train)
model3.weights_
```

```
array([0.02376651, 0.10343303, 0.0746895 , 0.08083832, 0.07263537,
       0.06226895, 0.0201518 , 0.08157259, 0.032706  , 0.03112282,
       0.21489253, 0.01559069, 0.04922161, 0.10995555, 0.01663003,
       0.01052472])
```
```
import skfolio; print(skfolio.__version__)
-->
0.4.2
```

EqualRisk portfolio returns weights to be 0.0625. The Portfolio HERC-CDaR-Ward-Kendal gives 0.21489258 (here for Novo Nordisk, which has had quite some fluctuations this year), (HERC-CDaR-Ward-Pearson gives 0.308940 with weight_constraints similarly set) 

I would like to have it and any other asset capped at max 0.1125 in this test , but seem to get almost double the weight of a single asset into portfolio.

I have tried to look at the herc.py code to understand if I am doing it wrong or whether there is something simple to correct in the code, but my python skills are not at adequate level to find the culprit. 

I can see from the code that there is intention to use the weight constraints after the HERC weights have been calculated, but I cannot understand the exact working of this. (No, I am not a programmer) 

Of course, I could post-process the portfolio weights in a separate code set and to calculate my own returns for the portfolio, but this would mean to abandoning the skfolio stats (portfolio.summary), which I find otherwise appealing.

Have I misunderstood the max_weights use or is use of weight constraints still under development or is there something wrong in the code?

Best regards,

Olli

**Comment by HugoDelatte** [comment]
Hello, thank you!

You are correct that HERC can be more unbalanced and concentrated compared to HRP.

As explained in this example: https://skfolio.org/auto_examples/5_clustering/plot_2_herc_cdar.html#linkage-methods, 
since HERC relies on the dendrogram structure (unlike HRP), the clustering can be significantly affected by the choice of linkage method and distance estimator.

The approach would be to first focus on the clustering (choosing the right linkage and distance), which will result in more balanced weights. As a last resort, you can indeed apply weight constraints.

Thanks for pointing out the bug. I'll investigate and push a fix this week.

**Comment by HugoDelatte** [comment]
@myllo issue is solved in v0.4.3.
You can find the details in the note section in:
https://skfolio.org/generated/skfolio.optimization.HierarchicalEqualRiskContribution.html

Thanks for spotting it.
